File: app/android/app/src/main/java/com/gonezo/multiplatform/audioextraction/infrastructure/resolution/NumberResolver.java
package com.gonezo.multiplatform.audioextraction.infrastructure.resolution;

import com.gonezo.multiplatform.audioextraction.domain.model.FieldCandidate;
import com.gonezo.multiplatform.audioextraction.domain.model.ResolvedField;
import com.gonezo.multiplatform.audioextraction.domain.schema.FieldSchema;
import java.math.BigDecimal;
import java.util.List;

final class NumberResolver implements FieldResolver {
  @Override
  public boolean supports(FieldSchema schema) {
    return "number".equals(schema.type()) || "integer".equals(schema.type());
  }

  @Override
  public ResolvedField resolve(String fieldName, FieldSchema schema, List<FieldCandidate> candidates) {
    FieldCandidate candidate = ResolverSupport.bestCandidate(candidates);
    if (candidate == null || candidate.rawValue() == null) {
      List<String> issues = schema.required() ? List.of("missing") : List.of();
      return new ResolvedField(null, 0D, List.of(), issues);
    }

    BigDecimal parsedNumber = parse(candidate.rawValue());
    if (parsedNumber == null) {
      return new ResolvedField(null, 0D, candidate.evidence(), List.of("invalid_format"));
    }

    if ("integer".equals(schema.type())) {
      return new ResolvedField(parsedNumber.longValue(), ResolverSupport.clamp(candidate.confidence()), candidate.evidence(), List.of());
    }

    return new ResolvedField(parsedNumber.doubleValue(), ResolverSupport.clamp(candidate.confidence()), candidate.evidence(), List.of());
  }

  private BigDecimal parse(Object rawValue) {
    if (rawValue instanceof Number numberValue) {
      return new BigDecimal(numberValue.toString());
    }

    if (!(rawValue instanceof String stringValue)) {
      return null;
    }

    String normalized = stringValue.trim();
    if (normalized.isEmpty()) {
      return null;
    }

    normalized = normalized.replace(" ", "").replace(",", ".");
    try {
      return new BigDecimal(normalized);
    } catch (NumberFormatException ex) {
      return null;
    }
  }
}

File: core/src/main/kotlin/com/gonezo/audioextraction/infrastructure/resolution/NumberResolver.kt
package com.gonezo.audioextraction.infrastructure.resolution

import com.gonezo.audioextraction.domain.model.FieldCandidate
import com.gonezo.audioextraction.domain.model.ResolvedField
import com.gonezo.audioextraction.domain.schema.FieldSchema
import java.math.BigDecimal

class NumberResolver : FieldResolver {
  override fun supports(schema: FieldSchema): Boolean = schema.type == "number" || schema.type == "integer"

  override fun resolve(fieldName: String, schema: FieldSchema, candidates: List<FieldCandidate>): ResolvedField {
    val candidate = ResolverSupport.bestCandidate(candidates)
    val rawValue = candidate?.rawValue
    if (candidate == null || rawValue == null) {
      val issues = if (schema.required) listOf("missing") else emptyList()
      return ResolvedField(null, 0.0, emptyList(), issues)
    }

    val parsedNumber = parse(rawValue)
    if (parsedNumber == null) {
      return ResolvedField(null, 0.0, candidate.evidence, listOf("invalid_format"))
    }

    val value: Any = if (schema.type == "integer") parsedNumber.toLong() else parsedNumber.toDouble()
    return ResolvedField(value, ResolverSupport.clamp(candidate.confidence), candidate.evidence, emptyList())
  }

  private fun parse(rawValue: Any): BigDecimal? {
    return when (rawValue) {
      is Number -> BigDecimal(rawValue.toString())
      is String -> rawValue.trim().replace(" ", "").replace(",", ".").takeIf { it.isNotEmpty() }?.toBigDecimalOrNull()
      else -> null
    }
  }
}
